File: app/src/main/java/com/divercity/app/features/profile/selectbirthdate/SelectBirthdayFragment.java
package com.divercity.app.features.profile.selectbirthdate;


import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.View;
import android.widget.Toast;

import com.divercity.app.R;
import com.divercity.app.core.base.BaseFragment;
import com.divercity.app.core.utils.Preconditions;
import com.divercity.app.databinding.FragmentSelectBirthdayBinding;
import com.divercity.app.features.dialogs.DatePickerDialogFragment;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

public class SelectBirthdayFragment extends BaseFragment implements DatePickerDialogFragment.DatePickerDialogListener {

    private static final int PROGRESS = 0;

    SelectBirthdayViewModel viewModel;
    FragmentSelectBirthdayBinding binding;

    @Inject
    Context context;

    public static SelectBirthdayFragment newInstance() {
        SelectBirthdayFragment fragment = new SelectBirthdayFragment();
        return fragment;
    }

    public SelectBirthdayFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_select_birthday;
    }

    @Override
    public ViewModel getViewModel() {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(SelectBirthdayViewModel.class);
        return viewModel;
    }

    @Override
    public int getBindingVariable() {
        return 0;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = (FragmentSelectBirthdayBinding) getViewDataBinding();
        setupHeader();
        subscribeToLiveData();
        binding.layDate.setOnClickListener(view1 -> {
            showDatePickerDialog();
        });
    }

    public void showDatePickerDialog() {
        DialogFragment f = DatePickerDialogFragment.newInstance();
        f.show(getChildFragmentManager(), null);
    }

    private void setupHeader(){
        binding.includeHeader.progressBar.setMax(100);
        binding.includeHeader.progressBar.setProgress(0);
        binding.includeHeader.progressBar.setProgressWithAnim(PROGRESS);
        binding.includeHeader.txtProgress.setText(PROGRESS + "%");
        binding.includeHeader.txtTitle.setText(R.string.select_your_birthday);

        binding.includeHeader.btnClose.setOnClickListener(view1 -> {

        });

        binding.includeHeader.btnSkip.setOnClickListener(view1 -> {

        });
    }

    private void subscribeToLiveData() {
        viewModel.dataUpdateUser.observe(this, response -> {
            Preconditions.checkNotNull(response);
            binding.includeLoading.setResource(response);
            switch (response.status){
                case ERROR:
                    Toast.makeText(getActivity(),response.message,Toast.LENGTH_SHORT).show();
                    break;
                case SUCCESS:
                    binding.txtDate.setText(response.data.getAttributes().getBirthdate());
                    break;
            }
        });
    }

    @Override
    public void onDateSetListener(String year, String month, String day) {
        viewModel.updateUserProfile(year,month,day);
    }
}

File: app/src/main/java/com/divercity/app/features/onboarding/selectbirthdate/SelectBirthdayFragment.kt
package com.divercity.app.features.onboarding.selectbirthdate

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.View
import android.widget.Toast
import com.divercity.app.R
import com.divercity.app.core.base.BaseFragment
import com.divercity.app.data.Status
import com.divercity.app.features.dialogs.DatePickerDialogFragment
import kotlinx.android.synthetic.main.fragment_select_birthday.*
import kotlinx.android.synthetic.main.view_header_profile.*

class SelectBirthdayFragment : BaseFragment(), DatePickerDialogFragment.DatePickerDialogListener {

    lateinit var viewModel: SelectBirthdayViewModel

    private var currentProgress: Int = 50

    private var isBirthdaySet: Boolean = false

    companion object {
        private const val PARAM_PROGRESS = "paramProgress"

        fun newInstance(progress: Int): SelectBirthdayFragment {
            val fragment = SelectBirthdayFragment()
            val arguments = Bundle()
            arguments.putInt(PARAM_PROGRESS, progress)
            fragment.arguments = arguments
            return fragment
        }
    }

    override fun layoutId(): Int = R.layout.fragment_select_birthday

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this, viewModelFactory)[SelectBirthdayViewModel::class.java]
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupHeader()
        subscribeToLiveData()

        lay_date.setOnClickListener { showDatePickerDialog() }

        btn_continue.setOnClickListener {
            if (isBirthdaySet)
                navigator.navigateToNextOnboarding(activity!!,
                        viewModel.accountType,
                        currentProgress,
                        true
                )
            else
                showToast("Set your birthday")
        }
    }

    private fun showToast(msg : String) {
        Toast.makeText(activity, msg, Toast.LENGTH_SHORT).show()
    }

    fun showDatePickerDialog() {
        val f = DatePickerDialogFragment.newInstance()
        f.show(childFragmentManager, null)
    }

    private fun setupHeader() {
        include_header.apply {
            currentProgress = arguments?.getInt(PARAM_PROGRESS) ?: 0

            progress_bar.apply {
                max = 100
                progress = 0
                setProgressWithAnim(currentProgress)
            }
            txt_title.setText(R.string.select_your_birthday)

            txt_progress.text = currentProgress.toString().plus("%")

            btn_close.setOnClickListener {
                navigator.navigateToHomeActivity(activity!!)
            }

            btn_skip.setOnClickListener {
                navigator.navigateToNextOnboarding(activity!!,
                        viewModel.accountType,
                        currentProgress,
                        false
                )
            }
        }
    }

    private fun subscribeToLiveData() {
        viewModel.updateUserProfileResponse.observe(this, Observer { school ->
            when (school?.status) {
                Status.LOADING -> showProgress()

                Status.ERROR -> {
                    hideProgress()
                    Toast.makeText(activity, school.message, Toast.LENGTH_SHORT).show()
                }
                Status.SUCCESS -> {
                    hideProgress()
                    txt_date.text = school.data?.attributes?.birthdate ?: "Error"
                    isBirthdaySet = true
                }
            }
        })
    }

    override fun onDateSetListener(year: String?, month: String?, day: String?) {
        viewModel.updateUserProfile(year, month, day)
    }
}
